demo/demo.js

```diff
@@ -28,27 +28,39 @@ CSS.paintWorklet.addModule('./ripple-worklet.js');
 
 if (!window.performance) window.performance = { now: Date.now.bind(Date) };
 
-if (!window.requestAnimationFrame) window.requestAnimationFrame = function(cb) { return setTimeout(doAnim, cb); };
+if (!window.requestAnimationFrame) {
+	window.requestAnimationFrame = function(cb) { return setTimeout(doAnim, 16, cb); };
+	window.cancelAnimationFrame = function(req) { clearTimeout(req); }
+}
 function doAnim(cb) { cb(performance.now()); }
 
-function ripple(evt) {
-	var button = this,
+function ripple(button, evt) {
+	var request,
 		rect = button.getBoundingClientRect(),
 		x = evt.clientX - rect.left,
 		y = evt.clientY - rect.top,
 		start = performance.now();
 	button.classList.add('animating');
-	requestAnimationFrame(function raf(now) {
+	request = requestAnimationFrame(function raf(now) {
 		var count = Math.floor(now - start);
 		button.style.cssText = '--ripple-x: ' + x + '; --ripple-y: ' + y + '; --animation-tick: ' + count + ';';
 		if (count > 1000) {
 			button.classList.remove('animating');
 			button.style.cssText = '--animation-tick: 0;';
 			return;
 		}
-		requestAnimationFrame(raf);
-	})
+		request = requestAnimationFrame(raf);
+	});
+	return function () {
```

Comment:
Could just return the rAF/timer ID here and avoid the cancellation proxy altogether.

Reply:
Actually that could let this be re-collapsed to a single method:

```js
addEventListener('click', function(e) {
  if (this.ripple) cancelAnimationFrame(this.ripple)

  ...
  this.ripple = requestAnimationFrame(rAF)
})
```

Reply:
`ripple()` can't return a request ID, because that ID changes after every `rAF` call.

We could indeed store the ID in a property on `button`, rather than keeping it inside a closure. However, I'm not really a fan of expando properties. 😛

For me, a "cleaner" solution would be to have a class instance per element, which handles both the `click` event and the animation loop. It'd be very similar to the current solution, except that all state moves from closures into class instances.